### PR TITLE
Time-stepper cleanup and modularization

### LINKED ISSUE #448: Logic for arguments to time-stepping related functions
Our time-stepping functions / kernels in `time_steppers.jl` are a bit chaotic with long argument lists, etc. To bring order to this world, it might be a good idea to adopt some kind of convention / preferred ordering to the function arguments. 

Almost all of the functions are mutating, so I think the first argument(s) should be the objects that are mutated by the function. This follows the [julia style guide recommendation](https://docs.julialang.org/en/v1/manual/style-guide/index.html#Write-functions-with-argument-ordering-similar-to-Julia-Base-1).

After that, the style guide doesn't offer guidance though it could make sense to put important objects for dispatch first (mostly `arch`). 

The rest of the objects fall into a few categories (I think):

* grid
* parameter structs
* `NamedTuple`s of fields / arrays / boundary conditions
* scalars

Perhaps we can use an ordering along the lines of

1. object being mutated
2. architecture (if used)
3. grid (if used)
4. scalars
5. parameter structs
6. `NamedTuple`s of stuff

I'm open to other patterns too.

### LINKED ISSUE #461: Don't be a type pirate for `NamedTuple`
We should nuke this function:

https://github.com/climate-machine/Oceananigans.jl/blob/b07dcf67f182283737b803a50a86c2ca5612fa75/src/utils.jl#L188

and come up with a different way to extract slices from our `NamedTuple`s, so that we do not commit the dreaded crime of type piracy.

cc @vchuravy

### LINKED ISSUE #485: Function to make velocity field non-divergent?
It might be nice to add a user-facing function that ensures the initial velocity field is non-divergent; eg `compute_w_from_continuity!(model)`. In the future we could also add options to compute `u` or `v`.

### PR DESCRIPTION
**Pull request by glwagner**
This PR moves the time stepping functionality into a module and refactors and orthogonalizes the time-stepping algorithm to make it easier to implement new time steppers. 

It also introduces a macro `@loop_xyz` that copy-pastes code to specify a 3D loop, reducing a lot of boiler plate.

There is also a new function `compute_w_from_continuity!(model)` that a user might employ to ensure their initial velocity field is non-divergent. 

Resolves #485 
Resolves #461 
Probably resolves #448

### CODE REVIEW

File: src/TimeSteppers/TimeSteppers.jl [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+    velocity_tendency_boundary_conditions = (u=model.boundary_conditions.tendency.u, 
+                                             v=model.boundary_conditions.tendency.v, 
+                                             w=model.boundary_conditions.tendency.w)
+
+    fill_halo_regions!(velocity_tendencies, velocity_tendency_boundary_conditions, model.architecture, model.grid)
+
+    @launch(device(model.architecture), config=launch_config(model.grid, 3), 
+            calculate_poisson_right_hand_side!(model.poisson_solver.storage, model.architecture, model.grid,
+                                               model.poisson_solver.bcs, velocities, tendencies, Δt))
```
Comment (ali-ramadhan) [inline src/TimeSteppers/TimeSteppers.jl]: Seems reasonable to split long `@launch` calls into multiple lines.

File: src/TimeSteppers/TimeSteppers.jl [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    for n in 1:Nt
+        time_step!(model, Δt; kwargs...)
+
+        [ time_to_run(model.clock, diag) && run_diagnostic(model, diag) for diag in values(model.diagnostics) ]
+        [ time_to_run(model.clock, out) && write_output(model, out) for out in values(model.output_writers) ]
+    end
+
+    return nothing
+end
+
```
Comment (ali-ramadhan) [inline src/TimeSteppers/TimeSteppers.jl]: Maybe add a banner like
```julia
####
#### Time-stepping kernels/function that are independent of the TimeStepper
####
```
as the individual steps are now split between multiple files.

File: src/models.jl [hunk 3]
```diff
@@ -58,7 +58,7 @@ function Model(;
              velocities = VelocityFields(architecture, grid),
               pressures = PressureFields(architecture, grid),
           diffusivities = TurbulentDiffusivities(architecture, grid, tracernames(tracers), closure),
-            timestepper = AdamsBashforthTimestepper(float_type, architecture, grid, tracernames(tracers), 0.125),
+            timestepper = :AdamsBashforth,
```
Comment (ali-ramadhan) [inline src/models.jl]: :+1:

File: src/utils.jl [hunk 4]
```diff
@@ -6,6 +6,23 @@ Adapt.adapt_structure(to, x::OffsetArray) = OffsetArray(adapt(to, parent(x)), x.
...
+        quote
+            @loop for $k in (1:$grid.Nz; (blockIdx().z - 1) * blockDim().z + threadIdx().z)
+                @loop for $j in (1:$grid.Ny; (blockIdx().y - 1) * blockDim().y + threadIdx().y)
+                    @loop for $i in (1:$grid.Nx; (blockIdx().x - 1) * blockDim().x + threadIdx().x)
+                        $expr
+                    end
+                end
+            end
+        end)
+end
```
Comment (ali-ramadhan) [inline src/utils.jl]: Love this macro!

File: src/TimeSteppers/TimeSteppers.jl (hunk 2 above)
Comment (glwagner) [reply src/TimeSteppers/TimeSteppers.jl]: Done.

File: src/TimeSteppers/TimeSteppers.jl (hunk 1 above)
Comment (glwagner) [reply src/TimeSteppers/TimeSteppers.jl]: It was an epiphany when I realized I can use parenthetical syntax for macros.

**Pull request merged**